typescript/src/monitor-printers.ts

```diff
@@ -0,0 +1,120 @@
+import { createWriteStream, existsSync } from 'fs';
+import {
+    mkdir, readdir, readFile, writeFile,
+} from 'fs/promises';
+import { DigitalFactoryDemo } from './digital-factory';
+import { prettyJSON, print } from './print';
+
+/**
+ * Contains the used fields from the Connect API cluster response.
+ */
+interface ClusterResponse {
+    cluster_id: string;
+    is_online: boolean;
+}
+
+/**
+ * Describes the comparison between two lists of clusters.
+ */
+interface Comparison {
+    dateTime: string;
+    appeared: string[];
+    gone: string[];
+    onlineCount: number;
+    offlineCount: number;
+}
+
+/**
+ * The options that may be given to this script.
+ */
+interface MonitorClustersOptions {
+    comparisonFile?: string;
+    statusDir?: string;
+    waitMS?: number;
+}
+
+/**
+ * Gets the previous cluster status from the status log directory. Ensures that the directory actually exists.
+ */
+async function getPreviousClusterStatus(statusDir: string): Promise<ClusterResponse[] | null> {
+    if (!existsSync(statusDir)) {
+        await mkdir(statusDir, { recursive: true });
+        return null;
+    }
+    const existingFiles = await readdir(statusDir);
+    const lastFile = existingFiles.filter((file) => file.endsWith('.json')).sort((a, b) => b.localeCompare(a))[0];
+    return lastFile && JSON.parse(await readFile(`${statusDir}/${lastFile}`, 'utf-8'));
+}
+
+/**
+ * Compares two lists of cluster statuses.
+ */
+function compareResults(previous: ClusterResponse[], current: ClusterResponse[], dateTime: Date): Comparison {
+    const getOnlineClusterIds = <T, K extends keyof any>(clusters: ClusterResponse[]): Set<string> => new Set<string>(
+        clusters.filter((cluster) => cluster.is_online).map((cluster) => cluster.cluster_id),
+    );
+    const previousOnlineIds = getOnlineClusterIds(previous);
+    const currentOnlineIds = getOnlineClusterIds(current);
+    return {
+        dateTime: dateTime.toISOString(),
+        appeared: [...currentOnlineIds].filter((id) => !previousOnlineIds.has(id)),
+        gone: [...previousOnlineIds].filter((id) => !currentOnlineIds.has(id)),
+        onlineCount: currentOnlineIds.size,
+        offlineCount: current.length - currentOnlineIds.size,
+    };
+}
+
+/**
+ * Creates a CSV line from the given comparison.
+ */
+function csvLine(comparison: Comparison, separator = ';'): string {
+    const fields: (keyof Comparison)[] = ['dateTime', 'appeared', 'gone', 'onlineCount', 'offlineCount'];
+    return fields.map((field) => comparison[field]).join(separator);
+}
+
+/**
+ * Gets all clusters from the API periodically, writing the response to a JSON file & the comparisons to a CSV file.
+ */
+async function monitorClusters({
+    comparisonFile = '../cluster-monitoring-logs/cluster-monitoring.csv',
+    waitMS = 60000,
+    statusDir = '../cluster-monitoring-logs',
+}: MonitorClustersOptions = {}): Promise<void> {
```

Comment:
Doesn't a default value for `MonitorClustersOptions` overwrite the defaults you specify before that?

Reply:
The ` = {}` part is the default parameter in case the function gets called without any parameters.
The `propName = ...` part is the default property in case the property isn't given.

For example:
![image](https://user-images.githubusercontent.com/2369243/145978578-871e047e-7077-4e9c-9b5d-5232e9f71bab.png)